src/media/mod.rs

```diff
@@ -0,0 +1,78 @@
+//!Media module
+//!
+//!Provides functionality to upload images, GIFs and videos using Twitter Media API
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::fmt::Write;
+use error;
+use error::Error::InvalidResponse;
+
+use rustc_serialize::json;
+
+use common::*;
+use links;
+use auth;
+
+///Represents media file that is uploaded on twitter.
+pub struct Media {
+    ///ID that can be used in API calls (e.g. attach to tweet).
+    pub id: u64,
+    ///Number of second the media can be used in other API calls.
+    pub expires_after: u64,
+}
+
+impl FromJson for Media {
+    fn from_json(input: &json::Json) -> Result<Self, error::Error> {
+        if !input.is_object() {
+            return Err(InvalidResponse("Tweet received json that wasn't an object", Some(input.to_string())));
+        }
+
+        field_present!(input, media_id);
+        field_present!(input, expires_after_secs);
+
+        Ok(Media {
+            id: try!(field(input, "media_id")),
+            expires_after: try!(field(input, "expires_after_secs"))
+        })
+    }
+}
+
+///Uploads image using old twitter API.
+///
+///The image should be raw binary with content of file.
+pub fn upload_image<'a, I: Into<Cow<'a, [u8]>>>(image: I, token: &auth::Token, handle: &'a Handle) -> FutureResponse<'a, Media> {
+    let mut params = HashMap::new();
+
+    let image = image.into();
+    add_param(&mut params, "media",
+              image.iter().fold(String::with_capacity(image.len()),
```

Comment:
I don't think this is the right way to encode this. For one, I wouldn't expect an image to be valid UTF-8, so the individual byte writes will fail at some point and make the `write!` call return an error, which will get unwrapped and cause a panic. The `media` parameter is probably not fit to be used with the `ParamList` anyway, since that expects to use strings everywhere, but the `media` wants to use the raw bytes. I would suggest putting this into `media_data` instead, and copying the code in [the auth module](https://github.com/QuietMisdreavus/twitter-rs/blob/3133af4ef802ac5f3859490a652053b2bbc352be/src/auth.rs#L414-L424) for encoding it as base64.

Reply:
Hm... I thought that there would be no issue to send raw bytes...
First of all bytes are [0..255] so they should be pretty much valid UTF-8 symbols,

Let me try to check something before going with base64. I don't really think there is a need for extra encoding to just send image. Seems wasteful.

Reply:
Actually, this won't panic, it will corrupt the image. Take the chunk encoded `0xDEADBEEF`. This code would take it one byte at a time, getting `0xDE`, `0xAD`, `0xBE`, `0xEF`. This code would then format each byte as a string of numbers (since it's going through `u8`'s Display impl), writing "222173190239" into the final String, which is definitely not the same as the input bytes.

Reply:
Ok, done as per your advise. Seems like no way to work it around :(